File: 프로그래머스/Level 2/기능개발.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds) {
    vector<int> answer;
    
    queue<pair<int, int>> q;
    
    for (int i = 0; i < progresses.size(); i++)
        q.push({progresses[i], i});
    
    int loc = 0;
    
    while (!q.empty()) {
        int cnt = 0;
        
        if (q.front().second == loc) {
        
            if (q.front().first >= 100) {
                while (!q.empty()) {
                    q.pop();
                    cnt++;
                    loc++;
                    if (q.front().first < 100)
                        break ;
                }
                answer.push_back(cnt);
            }
        }
        
        int idx = q.front().second;
        int tmp = q.front().first + speeds[idx];
        q.pop();
        q.push({tmp, idx});
    }
    
    return answer;
}

File: 프로그래머스/Level 2/기능개발.java
import java.util.*;

class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        ArrayList<Integer> list = new ArrayList<>();
        
        Queue<Integer> q_progresses = new LinkedList<>();
        Queue<Integer> q_speeds = new LinkedList<>();
        
        for (int i : progresses)
            q_progresses.add(i);
        
        for (int i : speeds) {
            q_speeds.add(i);
        }
        
        while (!q_progresses.isEmpty()) {
            int len = q_progresses.size();
            int cnt = 0;
            
            for (int i = 0; i < len; i++) {
                int progress = q_progresses.poll();
                int speed = q_speeds.poll();
                
                progress += speed;
                
                q_progresses.add(progress);
                q_speeds.add(speed);
            }
            
            for (int i = 0; i < len; i++) {
                if (q_progresses.peek() >= 100) {
                   cnt++;
                    q_progresses.poll();
                    q_speeds.poll();
                }
                else
                    break;
            }
            
            if (cnt > 0)
                list.add(cnt);
        }
        
        int[] answer = new int[list.size()];
        
        for (int i = 0; i < list.size(); i++)
            answer[i] = list.get(i);
        
        return answer;
    }
}
